### DISCUSSION TITLE
regex examples (negate regex)

### DISCUSSION THREAD

**Discussion by darked89** [BurntSushi/ripgrep#2226] [Q&A] [answered]
Hello,

I am trying to check if some CSV file contains bad values. The minimal example ```test.csv```:
```
123
123.0
123.0,456
123.4567890
123.4567890,bad_value
123.0\tany_not_a_number_coma_or_dot
```

doing:
```rg  -e  '[0-9]|,|\.' test.csv``` 

as expected shows all 0-9 numbers, dots and commas. But I am stuck trying to get either these two lines:
```
 123.4567890,bad_value
123.0\tany_not_a_number_coma_or_dot
```
or (better):
```
bad_value
\tany_not_a_number_coma_or_dot
```

Many thanks for your help.

Darek Kedra

**Comment by BurntSushi** [comment, marked as answer]
```
rg -e '[0-9]|,|\.' test.csv | rg -v -e bad_value -e not_a_number
```

Like that? Honestly, I'm having a lot of difficulty parsing your question. Could you please show the *entire* output that you want? Otherwise I'm just stuck trying to guess what you mean.

**Reply by darked89** [reply]
I am sorry for not being clear enough. 
The goal of the above ripgrep-ing:

Find any fishy chars, spaces including empty lines in a 2k columns and 20 up to 40k lines CSV files, or rather subset of these columns from #3 to #2k.
I use your ```xsv``` to exclude the first two columns.

The OK output:
whole lines in which there is at least one not(a_number or coma or dot).
I guess I should be able to turn on and off line numbering as needed as well as with -v just get the lines which are not containing any offending chars.

Before I was thinking about getting just the not(a_number or coma or dot) for each line, but this can get messy/hard to comprehend if there are multiple bad_chars per line.

In the end I will recheck the sanity of the input with ```polars```.

Is my question clear enough by now?

Thank you.

Darek Kedra

**Reply by BurntSushi** [reply]
So it sounds like you want ripgrep to print out any line that contains a character that _isn't_ a number, dot or comma? If so, then the strategy I would use is to write a regex that only matches those things and then invert it. So, `rg -v '^[0-9,.]+$'`.

**Reply by darked89** [reply]
This works perfect for me. Thank you and sorry for the long delay. Got swamped with stuff.

For some reason I am unable to mark this as an answer to the question. 

DK
